### PR TITLE
Iter9

### PR DESCRIPTION
**Pull request by ElizavetaFirst**

### CODE REVIEW

File: internal/agent/uploader/uploader.go [hunk 1]
```diff
@@ -94,3 +114,89 @@ func (u *Uploader) SendCounterMetrics(metrics map[string]int64) error {
 	}
 	return nil
 }
+
+func (u *Uploader) sendMetricsJSON(url string, metrics metrics.Metrics) error {
```
Comment (arttet) [inline internal/agent/uploader/uploader.go]: It seems this function should work with data as bytes

Comment (arttet) [inline internal/agent/uploader/uploader.go]:
```suggestion
func (u *Uploader) sendMetricsJSON(url string, metrics []byte) error {
```

File: internal/agent/uploader/uploader.go [hunk 2]
```diff
@@ -94,3 +114,89 @@ func (u *Uploader) SendCounterMetrics(metrics map[string]int64) error {
...
+		closeErr := reader.Close()
+		if closeErr != nil {
+			closeErr = errors.Wrap(closeErr, "Failed to close io.ReadCloser")
+			if err == nil {
+				err = closeErr
+			}
+		}
+	}()
+
+	switch resp.Header.Get("Content-Encoding") {
```
Comment (arttet) [inline internal/agent/uploader/uploader.go]: It would be better if it is in a middleware

File: internal/logger/logger.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	log, err = zap.NewProduction()
+	if err != nil {
+		fmt.Printf("can't initialize zap logger: %v", err)
+	}
+
+	return func(c *gin.Context) {
+		c.Set(logger, log)
+		c.Next()
+
+		if err := log.Sync(); err != nil && (!errors.Is(err, syscall.EBADF) && !errors.Is(err, syscall.ENOTTY)) {
```
Comment (arttet) [inline internal/logger/logger.go:27]:
```suggestion
		if err := log.Sync(); err != nil {
```

Comment (arttet) [reply internal/logger/logger.go:27]: Why do you need it?

**Review by arttet — CHANGES_REQUESTED** [review]
Please add the codecov report
https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/

File: internal/logger/logger.go (hunk 3 above)
Comment (ElizavetaFirst) [reply internal/logger/logger.go:27]: https://github.com/uber-go/zap/issues/328
проблема zap sync  на маке, но если не паниковать,то можно и без этого обойтись)

**Pull request closed without merging**